[JBPM-8635] Add prediction service examples

ruivieira opened the pull request:
Add example backends for the jBPM human task prediction service. Includes:

- PMML logistic regression and random forest
- SMILE random forest and naive Bayes

kie-ci4 commented:
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

ruivieira commented:
cc @mareknovotny @danielezonca @MarianMacik

mareknovotny commented:
ok to test

mareknovotny commented on jbpm-recommendation-pmml-logistic-regression/pom.xml:
```diff
@@ -0,0 +1,109 @@
+  <parent>
+    <groupId>org.jbpm</groupId>
+    <artifactId>jbpm</artifactId>
+    <version>7.27.0-SNAPSHOT</version>
```
this caused the CI failure, please update to 7.29.0-SNAPSHOT as while merging it needs to be for new modules in line with master branch version

ruivieira replied:
@mareknovotny thanks! I'll update the PR.

ruivieira commented:
@mareknovotny @danielezonca thank you for the review!

The PR was update with:

* Remove SMILE modules
* Remove dead code
* Add jPMML dep `scope` and `provided`
* Move PMML model to `test/resources`
* Add README.md with module documentation
* Update to 7.29-SNAPSHOT

mareknovotny commented:
there is something wrong with newly added maven modules and you probably did something wrong in rebasing from the master branch

```
22:09:57 [ERROR]   
22:09:57 [ERROR]   The project org.kie:kie-jpmml-integration:7.29.0-SNAPSHOT (/home/jenkins/workspace/KIE/master/pullrequest/kie-jpmml-integration-pullrequests/pom.xml) has 2 errors
22:09:57 [ERROR]     Child module /home/jenkins/workspace/KIE/master/pullrequest/kie-jpmml-integration-pullrequests/jbpm-recommendation-smile-random-forest of /home/jenkins/workspace/KIE/master/pullrequest/kie-jpmml-integration-pullrequests/pom.xml does not exist
22:09:57 [ERROR]     Child module /home/jenkins/workspace/KIE/master/pullrequest/kie-jpmml-integration-pullrequests/jbpm-recommendation-smile-naivebayes of /home/jenkins/workspace/KIE/master/pullrequest/kie-jpmml-integration-pullrequests/pom.xml does not exist
22:09:57 [ERROR] 
22:09:57 [ERROR] Re-run Maven using the -X switch to enable full debug logging.
22:09:57 [ERROR] 
22:09:57 [ERROR] For more information about the errors and possible solutions, please read the following articles:
22:09:57 [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
```

I get rebase conflict files on the first your commit if I create branch from current master branch. Please clear that error out.

MarianMacik commented on kie-dmn-jpmml/src/main/java/org/kie/dmn/jpmml/DMNjPMMLInvocationEvaluator.java:
```diff
@@ -89,7 +89,9 @@ public EvaluatorResult evaluate(DMNRuntimeEventManager eventManager, DMNResult d
-        return new EvaluatorResultImpl(result, ResultType.SUCCESS);
+
+        Object coercedResult = result.size() > 1 ? result : result.values().iterator().next();
+        return new EvaluatorResultImpl(coercedResult, ResultType.SUCCESS);
```
Is this a part of the PR?

ruivieira replied:
From Kie `master` (merged instead of rebased).

MarianMacik commented on kie-dmn-jpmml/src/test/java/org/kie/dmn/jpmml/DMNInvokingjPMMLTest.java:
```diff
@@ -115,6 +118,33 @@ public void testDummyInteger() {
+        assertThat(resultOfHardcoded, hasEntry("Predicted_Species", "virginica"));
+        assertThat(resultOfHardcoded, hasKey("Predicted_Petal.Width"));
+        assertThat((BigDecimal) resultOfHardcoded.get("Predicted_Petal.Width"), is(closeTo(new BigDecimal("1.9333333333333336"), new BigDecimal("0.1"))));
+        // no special interest to check the other output fields as the above are the user-facing most interesting ones.
```
Is this a part of the PR?

ruivieira replied:
From Kie `master` (merged instead of rebased).

MarianMacik commented on pom.xml:
```diff
@@ -5,14 +5,18 @@
   <modules>
     <module>kie-dmn-jpmml</module>
+    <module>jbpm-recommendation-smile-random-forest</module>
+    <module>jbpm-recommendation-smile-naivebayes</module>
```
`smile` modules seem to be redundant?

ruivieira replied:
Removed now.

ruivieira commented on jbpm-recommendation-pmml-logistic-regression/src/main/java/org/jbpm/prediction/pmml/AbstractPMMLBackend.java:
```diff
@@ -0,0 +1,109 @@
+                    .build();
+            _evalutator.verify();
+        } catch (IOException | SAXException | JAXBException e) {
+            e.printStackTrace();
```
Exception no longer suppressed.

ruivieira commented on jbpm-recommendation-pmml-logistic-regression/src/main/java/org/jbpm/prediction/pmml/PMMLLogisticRegression.java:
```diff
@@ -0,0 +1,187 @@
+            } else {
+                throw new FileNotFoundException("Could not find the property file 'inputs.properties' in the classpath.");
+            }
+
```
Property file loading refactored into separate method.

ruivieira commented on jbpm-recommendation-pmml-logistic-regression/src/test/java/org/jbpm/prediction/pmml/PMMLLogisticRegressionPredictionServiceProcessTest.java:
```diff
@@ -0,0 +1,111 @@
+
+    /**
+     * The PMML model was trained with 90% of "john" outcomes as "false"
+     * Expect confidence > 0.9 (90.0%) and "approved" to be false.
```
The PMML was trained using Python’s scikit-learn for this specific example (_i.e._ with John and Mary as the “actors” and with categories coded as `John=0` and `Mary=1`. I could include the model training script but I thought it would be outside the scope of this repo.

MarianMacik replied:
It's ok, I just wanted to be sure how it works. Thanks.

ruivieira commented on jbpm-recommendation-pmml-random-forest/src/main/java/org/jbpm/prediction/pmml/PMMLRandomForest.java:
```diff
@@ -0,0 +1,193 @@
+        String predictionStr;
+
+        Double prediction = (Double) result.get(outcomeFeatureName);
+        double confidence = Math.max(Math.abs(0.0 - prediction), Math.abs(1.0 - prediction));
```
Good spot! :)

ruivieira commented on pom.xml:
```diff
@@ -5,14 +5,18 @@
     <groupId>org.kie</groupId>
     <artifactId>kie-parent</artifactId>
-    <version>7.27.0-SNAPSHOT</version>
+    <version>7.29.0-SNAPSHOT</version>
```
Kie `master` merged instead of rebased.

ruivieira commented:
@MarianMacik @mareknovotny thank you for the review! I've addressed the comments individually, but I'll summarise here:

* Commits not by myself, present due to a merge with kie's `master` instead of a rebase
* Version bumped to `7.30-SNAPSHOT` in the modules
* Exceptions no longer suppressed when PMML fails to load the model
* `Math.abs(0.0 - prediction)` simplified to `prediction`
* SMILE references deleted from parent POM

mareknovotny commented:
@ruivieira I see that and generally you should not pull updates from base branch to do that, but rebase with forwarded changes. This makes hard to review here the only changes you did here.

MarianMacik commented on kie-dmn-jpmml/src/test/java/org/kie/dmn/jpmml/DMNInvokingjPMMLTest.java:
```diff
@@ -115,6 +118,33 @@ public void testDummyInteger() {
+    }
+
+    @Test
+    public void testMultipleOutputs() {
```
BTW I think these changes in kie-dmn-jpmml module are already on master so should be excluded from this PR.

mswiderski commented:
@ruivieira looks like the version in pom.xml should be now changed to 7.32.0-SNAPSHOT, could you please make this change and then we merge it

mareknovotny commented on jbpm-recommendation-pmml-logistic-regression/pom.xml:
```diff
@@ -0,0 +1,113 @@
+  <parent>
+    <groupId>org.jbpm</groupId>
+    <artifactId>jbpm</artifactId>
+    <version>7.30.0-SNAPSHOT</version>
```
rebase the version to master one which is 7.32.0-SNAPSHOT

mareknovotny replied:
Why is here parent outside of this repository ?

mareknovotny commented on pom.xml:
```diff
@@ -5,14 +5,16 @@
     <groupId>org.kie</groupId>
     <artifactId>kie-parent</artifactId>
-    <version>7.27.0-SNAPSHOT</version>
+    <version>7.30.0-SNAPSHOT</version>
```
rebase the version to master one which is 7.32.0-SNAPSHOT
